lib/apis/v3/node.go

```diff
@@ -98,6 +103,27 @@ type NodeWireguardSpec struct {
 	InterfaceIPv4Address string `json:"interfaceIPv4Address,omitempty" validate:"omitempty,ipv4"`
 }
 
+// NodeAddressType represents a node address type as a string
+type NodeAddressType string
+
+const (
+	NodeAddressBGPIP             NodeAddressType = "BGPIP"
+	NodeAddressVXLANTunnelIP     NodeAddressType = "VXLANTunnelIP"
+	NodeAddressVXLANMAC          NodeAddressType = "VXLANMAC"
+	NodeAddressIPIPTunnelIP      NodeAddressType = "IPIPTunnelIP"
+	NodeAddressWireguardTunnelIP NodeAddressType = "WireguardTunnelIP"
+	NodeAddressKubeInternalIP    NodeAddressType = "KubeInternalIP"
```

Comment:
If we do actually need to track the source of the address, my suggestion would be to extend the NodeAddress struct rather than try to put two bits of information (type and source) into a single field. 

e.g.,

```yaml
- address: 1.2.3.4/32
  type: InternalIP
  source: Kubernetes

```

Reply:
Not sure about the source. It is more about what is the role of the address. What is a meaning of an "internal IP"? Perhaps in the module that uses it, which is always executed in kubernetes, the meaning is well defined and the const should perhaps defined there.

If we extend the struct we loose the nice compatibility properties.

Reply:
> If we extend the struct we loose the nice compatibility properties.

Hm, say more about this. What compatibility properties are you referring to? 

>  It is more about what is the role of the address. What is a meaning of an "internal IP"?

Yes, I agree. And IIUC, the role of the address is not "An internal IP that we learned from Kubenrnetes" but "An internal IP", right?

An internal IP is one that can be used to access the node from within the cluster, I think.

Reply:
> > If we extend the struct we loose the nice compatibility properties.
>
> Hm, say more about this. What compatibility properties are you referring to?

If we wanted to add an address that does not have a source, but some other property (well, for instance a version like IPv4/6), then some addresses use one field, others other or all fields etc. A union would be nice. Imo an `int` would serve us well, but the string also carries the human readable version, which is nice :) Feels like structuring got us here.

> > It is more about what is the role of the address. What is a meaning of an "internal IP"?
> 
> Yes, I agree. And IIUC, the role of the address is not "An internal IP that we learned from Kubenrnetes" but "An internal IP", right?

It is an IP assigned by kubernetes and it is internal to a kubernetes cluster. Technically the BPG address (which we learn by a range of methods) is usually assigned by the same source, has a different role it seems to me. But KubeInternalIPv4 feels a bit too much. I guess the cleanest / future proof might be:

```
type NodeAddress struct {
    Type NodeAddressType // IP, MAC, DNS, HostName etc.
    Annotations map[string]string
}

k8sInternalIPv4 := {
    Address: "1.2.3.4",
    Type: NodeAddressTypeIP,
    Annotations: map[string]string{"version":"v4", "source":"kubernetes-internal"},
}

eth0IP := {
    Address: "1.2.3.4",
    Type: NodeAddressTypeIP,
    Annotations: map[string]string{"version":"v4", "source":"netlink", "iface":"eth0"},
}
```

It is a bit more heavyweight and uncessary.

If we did the deduplication of address here, we could merge the annotations (what about source?) but that would not work well with etcd, so better to leave it to the user...